src/terms/nonlocal.jl

```diff
@@ -178,22 +178,34 @@ end
 """
 Build form factors (Fourier transforms of projectors) for an atom centered at 0.
 """
-function build_form_factors(psp, qs)
-    qnorms = norm.(qs)
-    T = real(eltype(qnorms))
-    # Compute position-independent form factors
-    form_factors = zeros(Complex{T}, length(qs), count_n_proj(psp))
-    count = 1
-    for l in 0:psp.lmax, m in -l:l
-        prefac_lm = im^l .* ylm_real.(l, m, qs)
-        n_proj_l = size(psp.h[l + 1], 1)
+function build_form_factors(psp, G_plus_ks)
+    T = real(eltype(first(G_plus_ks)))
+
+    n_proj_max = maximum(l -> count_n_proj_radial(psp, l), 0:psp.lmax; init=0)
+    radials = IdDict{T,Matrix{T}}()
+    for Gpk in G_plus_ks
+        q = norm(Gpk)
+        if !haskey(radials, q)
+            radials_q = Matrix{T}(undef, n_proj_max, psp.lmax + 1)
+            for l in 0:psp.lmax, iproj_l in 1:count_n_proj_radial(psp, l)
+                radials_q[iproj_l, l+1] = eval_psp_projector_fourier(psp, iproj_l, l, q)
+            end
+            radials[q] = radials_q
+        end
+    end
 
-        for iproj in 1:n_proj_l
-            radial_il = eval_psp_projector_fourier.(psp, iproj, l, qnorms)
-            form_factors[:, count] = prefac_lm .* radial_il
-            count += 1
+    form_factors = Matrix{Complex{T}}(undef, length(G_plus_ks), count_n_proj(psp))
+    for (iGpk, Gpk) in enumerate(G_plus_ks)
+        radials_q = radials[norm(Gpk)]
+        count = 1
+        for l in 0:psp.lmax, m in -l:l
+            angular_Glm = im^l * ylm_real(l, m, Gpk)
+            for iproj_l in 1:count_n_proj_radial(psp, l)
+                form_factors[iGpk, count] = angular_Glm * radials_q[iproj_l, l+1]
+                count += 1
+            end
         end
+        @assert count == count_n_proj(psp) + 1
     end
-    @assert count == count_n_proj(psp) + 1
     form_factors
-end
+end
```

Comment:
newline at end of file (here and other places)

Reply:
Using JuliaFormatter as a pre-commit hook could be very nice for this and other formatting issues.

Reply:
I would advertise :-) a bit my thin wrapper of `JuliaFormatter.jl` as a pre-commit hook, it ensures each commit is formatted
https://github.com/qiaojunfeng/pre-commit-julia-format

Reply:
I might change my mind just to get rid of discussions around style :-) there's definitely some things I could get behind (newlines at end of files, required semicolon before kwargs, these two spaces before # Michael loves so much, etc) but I'm super reluctant about automatic spacing and line breaks